### PR TITLE
[Compatibility] Adding EXPIRETIME and PEXPIRETIME command

### PR DESCRIPTION
**Pull request by Vijay-Nirmal**
Adding EXPIRETIME and EXPIRETIME command to garnet

- [x] Add EXPIRETIME and EXPIRETIME command
- [x] Add Test cases
- [x] Add documentation

Open Question:
1. For Object type, Should I create a single GarnetObjectType called `Expiration` instead of `EXPIRETIME` and `PEXPIRETIME`? The advantage will be that it will be a single generic implementation that can be reused for others in the future if required. The disadvantage is that there will be a slight performance penalty but no additional memory needed since values of `Expiration`, `EXPIRETIME` and `PEXPIRETIME` are `long` type, so we can do in-place override the value after recomputing the required value.

### CODE REVIEW

**Comment by TalZaccai** [comment]
> Adding EXPIRETIME and EXPIRETIME command to garnet
> 
> * [x]  Add EXPIRETIME and EXPIRETIME command
> * [x]  Add Test cases
> * [x]  Add documentation
> 
> Open Question:
> 
> 1. For Object type, Should I create a single GarnetObjectType called `Expiration` instead of `EXPIRETIME` and `PEXPIRETIME`? The advantage will be that it will be a single generic implementation that can be reused for others in the future if required. The disadvantage is that there will be a slight performance penalty but no additional memory needed since values of `Expiration`, `EXPIRETIME` and `PEXPIRETIME` are `long` type, so we can do in-place override the value after recomputing the required value.

There is already a GarnetObjectType.Expire, why not use that one?

File: libs/server/Storage/Session/MainStore/MainStoreOps.cs
Comment (TalZaccai) [inline libs/server/Storage/Session/MainStore/MainStoreOps.cs:320]: Why not simply call TTL and convert it to unix-time?

**Review by TalZaccai — CHANGES_REQUESTED** [review]
Thanks for your contribution! Added a couple of comments.

File: libs/server/Storage/Session/MainStore/MainStoreOps.cs
Comment (Vijay-Nirmal) [reply libs/server/Storage/Session/MainStore/MainStoreOps.cs:320]: That was my first implementation, and then I did this implementation and did a microbenchmark to compare the performance. I found this method is much faster than calling editing TTL and converting to Unix time. I don't have the number handy now but it was like 60% difference. TTL + Convert took about 15 ns and this approach was taking about 9 ns. I know its still few `ns`. @TalZaccai If you are fine with this performance difference then I can change it to TTL + Convert logic. Your call

File: libs/server/Objects/Types/GarnetObjectType.cs
Comment (Vijay-Nirmal) [reply libs/server/Objects/Types/GarnetObjectType.cs:35]: As mentioned in the other comment if we are going to use TTL + Convert logic then this is not need. I will wait for you confirmation before changing as I mentioned about minor perf difference.

File: libs/server/Storage/Session/MainStore/MainStoreOps.cs
Comment (TalZaccai) [inline libs/server/Storage/Session/MainStore/MainStoreOps.cs:350]: FYI - You can't set both cmd and type since they're written to the same area of memory, use only type in this case.

File: libs/server/Storage/Session/MainStore/MainStoreOps.cs
Comment (TalZaccai) [reply libs/server/Storage/Session/MainStore/MainStoreOps.cs:320]: Interesting, I wonder why it would take more time, as this would still do the conversion, only in a different area of the code. 

@badrishc do you have a preference here?

File: libs/server/Storage/Session/MainStore/MainStoreOps.cs
Comment (Vijay-Nirmal) [reply libs/server/Storage/Session/MainStore/MainStoreOps.cs:350]: Thanks for the info, I didn't notice the `FieldOffset` set to `0` for both the fields, Will have the changes required

File: libs/server/Storage/Session/MainStore/MainStoreOps.cs
Comment (Vijay-Nirmal) [reply libs/server/Storage/Session/MainStore/MainStoreOps.cs:320]: Reason for the performance difference is because TTL will write its response in `SpanByteAndMemory` and change its available length. Since UnixTimestamp is always bigger than TTL or PTTL value we are override in the existing space in `SpanByteAndMemory` which makes the code to `Rent` a new memory for the response.

One way I though of overcome this is by resetting the length of `SpanByteAndMemory` to initial after getting the response from TTL/PTTL but not sure how safe it is. I think it would be safe as I would assume those a pinned memory and won't be disposed till we right the response back from the server. If you are fine with doing this then I can use TTL/PTTL itself or if there is a better way, let me know I can implement in that manner.

Comment (Vijay-Nirmal) [reply libs/server/Storage/Session/MainStore/MainStoreOps.cs:320]: @TalZaccai @badrishc Any update? I can make the suggested changes if you guys let me know what you prefer.

File: libs/server/Storage/Session/MainStore/MainStoreOps.cs
Comment (Vijay-Nirmal) [reply libs/server/Storage/Session/MainStore/MainStoreOps.cs]: Fixed it

File: libs/server/Storage/Functions/ObjectStore/ReadMethods.cs
Comment (badrishc) [inline libs/server/Storage/Functions/ObjectStore/ReadMethods.cs:40]: Can create a single switch with all variants of TTL and ExpireTime, to consolidate.

File: libs/server/Storage/Session/MainStore/MainStoreOps.cs
Comment (badrishc) [reply libs/server/Storage/Session/MainStore/MainStoreOps.cs:320]: Since TTL writes directly to the network, it should be fine to have a separate ExpireTime that writes the time directly to the network buffer as this PR does.

**Review by badrishc — APPROVED** [review]

File: libs/server/Storage/Functions/ObjectStore/ReadMethods.cs
Comment (Vijay-Nirmal) [reply libs/server/Storage/Functions/ObjectStore/ReadMethods.cs]: Added `switch` for the entire type so that in future if there is more `GarnetObjectType` that needs to be handled in the ObjectSession layer itself, we can use the same switch

**Comment by Vijay-Nirmal** [comment]
@badrishc @TalZaccai Fixed the pending review comments

**Review by TalZaccai — APPROVED** [review]
Thank you! Please resolve conflicts and I will merge this in :)

**Comment by Vijay-Nirmal** [comment]
@TalZaccai Merged the latest

**Review by TalZaccai — APPROVED** [review]

**Pull request merged**
